samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -65,12 +66,10 @@
   private volatile String ephemeralPath = null;
   private final ZkKeyBuilder keyBuilder;
   private final int connectionTimeoutMs;
-  private ZkJobCoordinatorMetrics metrics;
+  private final ZkJobCoordinatorMetrics metrics;
 
   public ZkUtils(ZkKeyBuilder zkKeyBuilder, ZkClient zkClient, int connectionTimeoutMs) {
-    this.keyBuilder = zkKeyBuilder;
-    this.connectionTimeoutMs = connectionTimeoutMs;
-    this.zkClient = zkClient;
+    this (zkKeyBuilder, zkClient, connectionTimeoutMs, new ZkJobCoordinatorMetrics(new MetricsRegistryMap()));
```

Comment:
Please check my changes and verify if it makes sense. Though I partly agree with you on approach, lets at-least pass ZkUtilsMetrics into ZkUtils constructor. Passing in ZkJobCoordinatorMetrics into ZkUtils constructor looks bizarre to me.
 
I'm bringing in new Reporters in `ZkCoordinationServiceFactory `, not sure how much resource footprint it will create.